### PR TITLE
Refactoring flexible CNV variant loader

### PR DESCRIPTION
**Pull request by lchorbadjiev**
This work was motivated by the need to make CNV variant loader easier to understand and test.

A flexible variant loader function was introduced to load column-oriented files and transform each line in such a file into a dictionary.

The transformation of the line into a dictionary is done by the function after that a series of transformers are applied sequentially to each record to convert it into the desired form.

The flexible variant loader function supports applying a series of filters to each record. Only records that pass all filters are emitted.

A set of transformers and helper functions are defined to support the loading of CNV variants.

### CODE REVIEW

File: dae/dae/backends/cnv/tests/test_configure_cnv_loader.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+    header = ["personId", "location", "variant", "bestState", "extra"]
+    transformers = []
+
+    with pytest.raises(ValueError):
+        _configure_cnv_best_state(
+            header, transformers,
+            families, gpf_instance_2013.reference_genome,
+            cnv_person_id="personId",
+            cnv_best_state="bestSTate",
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_configure_cnv_loader.py:134]: should this "bestSTate" string be misspelled?

File: dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    return families
+
+
+@pytest.fixture
+def genome(gpf_instance_2013):
+    return gpf_instance_2013.reference_genome
+
+
+@pytest.mark.parametrize(
+    # params = (cnv_family_id, cnv_location, cnv_variant_type, cnv_best_state)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py:37]: commented-out code

File: dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    rec = result[0]
+    assert rec["chrom"] == "1"
+    assert rec["pos"] == 1590681
+    assert rec["pos_end"] == 1628197
+    assert rec["variant_type"] == Allele.Type.large_duplication
+    assert mat2str(rec["best_state"]) == "2221/0001"
+
+
+@pytest.mark.parametrize(
+    # params = (cnv_person_id, cnv_chrom, cnv_start, cnv_end, cnv_variant_type)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py:82]: commented-out code

File: dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    with pytest.raises(ValueError) as ex:
+        next(
+            flexible_cnv_loader(
+                content,
+                families,
+                genome,
+                [],
+                **params)
+            )
+    print(ex)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py:247]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+           "cnv_location": "location",
+           "cnv_variant_type": "variant",
+           "cnv_plus_values": ["GAIN"],
+           "cnv_minus_values": ["LOSS"],
+           "cnv_transmission_type": transmission_type,
+        }
+    )
+
+    variants = list(loader.full_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:61]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+           "cnv_location": "location",
+           "cnv_variant_type": "variant",
+           "cnv_plus_values": ["GAIN"],
+           "cnv_minus_values": ["LOSS"],
+           "cnv_transmission_type": "transmitted",
+        }
+    )
+
+    variants = list(loader.full_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:92]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+           "cnv_variant_type": "variant",
+           "cnv_plus_values": ["GAIN"],
+           "cnv_minus_values": ["LOSS"],
+           "cnv_transmission_type": "transmitted",
+           "add_chrom_prefix": add_chrom_prefix,
+        }
+    )
+
+    variants = list(loader.full_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:129]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    assert len(variants) == 2
+
+    loader.reset_regions(region)
+    variants = list(loader.family_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:134]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+           "cnv_variant_type": "variant",
+           "cnv_plus_values": ["GAIN"],
+           "cnv_minus_values": ["LOSS"],
+           "cnv_transmission_type": "transmitted",
+           "add_chrom_prefix": add_chrom_prefix,
+        }
+    )
+
+    variants = list(loader.family_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:168]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+           "cnv_variant_type": "variant",
+           "cnv_plus_values": ["GAIN"],
+           "cnv_minus_values": ["LOSS"],
+           "cnv_transmission_type": "transmitted",
+           "del_chrom_prefix": "chr",
+        }
+    )
+
+    variants = list(loader.full_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:206]: print

File: dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    assert len(variants) == 2
+
+    loader.reset_regions(region)
+    variants = list(loader.family_variants_iterator())
+    print(variants)
```
Comment (iordanivanov) [inline dae/dae/backends/cnv/tests/test_cnv_abn_canvas_loader.py:211]: print

File: dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py (hunk 2 above)
Comment (lchorbadjiev) [reply dae/dae/backends/cnv/tests/test_flexible_cnv_loader.py]: this is not a code, but a comment for the structure of the params tuple;

File: dae/dae/backends/cnv/tests/test_configure_cnv_loader.py (hunk 1 above)
Comment (lchorbadjiev) [reply dae/dae/backends/cnv/tests/test_configure_cnv_loader.py:134]: this test intentional error in the name of best_state... expects to raise ValueError

**Review by IvoTod — APPROVED** [review]

**Pull request merged**